src/ICS20Transfer.sol

```diff
@@ -148,18 +145,47 @@ contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, Reentr
     /// @inheritdoc IIBCApp
     function onTimeoutPacket(OnTimeoutPacketCallback calldata msg_) external onlyOwner nonReentrant {
         ICS20Lib.PacketDataJSON memory packetData = ICS20Lib.unmarshalJSON(msg_.packet.data);
-        (address erc20Address,) = getSendERC20AddressAndSource(msg_.packet, packetData);
-        _refundTokens(packetData, erc20Address);
+        _refundTokens(msg_.packet, packetData);
 
         emit ICS20Timeout(packetData);
     }
 
+    /// @notice Get the escrow address for a channel ID
+    /// @param channelId The channel ID
+    /// @return The address of the escrow contract
+    function _getEscrowAddress(string memory channelId) internal returns (address) {
+        address ics20EscrowAddress = address(_channelEscrowContracts[channelId]);
+        if (ics20EscrowAddress == address(0)) {
+            // nothing exists, so we create new contract and register it in the mapping
+            ICS20Escrow ics20EscrowContract = new ICS20Escrow(IICS20Transfer(address(this)));
+            _channelEscrowContracts[channelId] = ics20EscrowContract;
+            ics20EscrowAddress = address(ics20EscrowContract);
+        }
+
+        return ics20EscrowAddress;
+    }
+
     /// @notice Refund the tokens to the sender
+    /// @param packet The ICS26 packet
     /// @param packetData The packet data
-    /// @param erc20Address The address of the ERC20 contract
-    function _refundTokens(ICS20Lib.PacketDataJSON memory packetData, address erc20Address) private {
+    function _refundTokens(
+        IICS26RouterMsgs.Packet calldata packet,
+        ICS20Lib.PacketDataJSON memory packetData
+    )
+        private
+    {
         address refundee = ICS20Lib.mustHexStringToAddress(packetData.sender);
-        IERC20(erc20Address).safeTransfer(refundee, packetData.amount);
+        (address erc20Address, bool senderChainIsSource) = getSendERC20AddressAndSource(packet, packetData);
+
+        if (senderChainIsSource) {
+            address ics20EscrowAddress = _getEscrowAddress(packet.sourceChannel);
+
+            // transfer the tokens back to the sender from the escrow address
+            _transferFrom(ics20EscrowAddress, refundee, erc20Address, packetData.amount);
+        } else {
+            // tokens had been burned when they were sent, so we must mint them back
+            IBCERC20(erc20Address).mint(refundee, packetData.amount);
+        }
```

Comment:
When refunding tokens: if the chain is source of the tokens, then unscrew them to the sender account; otherwise mint vouchers.